### PR TITLE
Fix races from indexing deletions (#611)

### PR DESCRIPTION
**Pull request by jessebrennan**
Now, deletions count as an entirely new contribution. When aggregating, the version is skipped if a deletion contribution exists for it. This means that the order for receiving a deletion and an addition notification doesn't matter

### CODE REVIEW

**Comment by jessebrennan** [comment]
@hannes-ucsc the one suggestion that I did not yet add was a refactor to add a `BundleInfo` class that would make some function signatures more compact. I started this, but it ran into some complicated issues so I've put it off. Things seem good enough in the current state anyway (though you might disagree).

File: src/azul/indexer.py [hunk 1]
```diff
@@ -246,39 +246,26 @@ def _read_contributions(self, tallies: Tallies) -> List[Contribution]:
...
-                pass
-            else:
-                contributions_by_entity[contribution.entity].append(contribution)
+            tallies[contribution.entity] += 1
+
+        # For each entity and bundle, find the most recent contribution that is not a deletion
+        contributions_by_entity: Mapping[EntityReference, List[Contribution]] = defaultdict(list)
+        for (entity, bundle_uuid), contributions in contributions_by_bundle.items():
+            contributions = sorted(contributions, key=attrgetter('bundle_version', 'bundle_deleted'), reverse=True)
+            for bundle_version, group in groupby(contributions, key=attrgetter('bundle_version')):
```
Comment (jessebrennan) [inline src/azul/indexer.py]: @hannes-ucsc this looks great! I wish I'd known about groupby. I'm a little envious that you wrote this code ;)

File: src/azul/indexer.py [hunk 2]
```diff
@@ -119,9 +126,11 @@ def _get_bundle(self, bundle_uuid, bundle_version):
         assert _ == bundle_version
         return manifest, metadata_files
 
-    def _add_test_modifications(self, manifest, metadata_files, dss_notification):
+    def _add_test_modifications(self, manifest, metadata_files, dss_notification, default):
```
Comment (hannes-ucsc) [inline src/azul/indexer.py]: `default` is not very descriptive. But see comment at call site.
```suggestion
    def _add_test_modifications(self, bundle_uuid, manifest, metadata_files, dss_notification):
```

File: src/azul/indexer.py [hunk 3]
```diff
@@ -257,30 +264,40 @@ def _aggregate(self, contributions: List[Contribution]) -> List[Aggregate]:
...
+        contributions = self.transform(dss_notification, delete=True)
         # FIXME: these are all modified contributions, not new ones. This also happens when we reindex without
         # deleting the indices first. The tallies refer to number of updated or added contributions but we treat them
         # as if they are all new when we estimate the number of contributions per bundle.
         # https://github.com/DataBiosphere/azul/issues/610
         tallies = self.contribute(writer, contributions)
         self.aggregate(writer, tallies)
 
-    def transform_deletion(self, dss_notification: JSON) -> List[Contribution]:
+    def reflective_transform_delete(self, dss_notification: JSON) -> List[Contribution]:
```
Comment (hannes-ucsc) [inline src/azul/indexer.py]: Let's just delete it.

File: src/azul/indexer.py [hunk 4]
```diff
@@ -72,44 +75,48 @@ def index(self, writer: 'IndexWriter', dss_notification: JSON) -> None:
...
+            return self.reflective_transform_delete(dss_notification)
         bundle_uuid = dss_notification['match']['bundle_uuid']
         bundle_version = dss_notification['match']['bundle_version']
         manifest, metadata_files = self._get_bundle(bundle_uuid, bundle_version)
         # If indexing a test bundle we want to change the uuid so that we can delete the bundle after
-        test_bundle_uuid = self._add_test_modifications(manifest, metadata_files, dss_notification)
-        if test_bundle_uuid:
-            bundle_uuid = test_bundle_uuid
-
+        bundle_uuid = self._add_test_modifications(manifest, metadata_files, dss_notification, bundle_uuid)
```
Comment (hannes-ucsc) [inline src/azul/indexer.py]:
```suggestion
        bundle_uuid = self._add_test_modifications(bundle_uuid, manifest, metadata_files, dss_notification)
```

File: src/azul/indexer.py [hunk 5]
```diff
@@ -72,44 +75,48 @@ def index(self, writer: 'IndexWriter', dss_notification: JSON) -> None:
...
         tallies = self.contribute(writer, contributions)
         self.aggregate(writer, tallies)
 
-    def transform(self, dss_notification: JSON) -> List[Contribution]:
+    def transform(self, dss_notification: JSON, delete: bool) -> List[Contribution]:
         """
         Transform the metadata in the bundle referenced by the given notification into a list of contributions to
         documents, each document representing one metadata entity in the index.
         """
+        if delete and False:
```
Comment (hannes-ucsc) [inline src/azul/indexer.py]: Let's just delete reflective_transform_delete.

File: src/azul/transformer.py [hunk 6]
```diff
@@ -176,6 +198,7 @@ def transform(self,
 
         :param uuid: The UUID of the bundle to create documents for
         :param version: The version of the bundle to create documents for
+        :param deleted: Whether or not the bundle being indexed is deleted
```
Comment (hannes-ucsc) [inline src/azul/transformer.py]:
```suggestion
        :param deleted: Whether or not the bundle being indexed has been deleted
```

???

**Comment by hannes-ucsc** [comment]
And could you please rebase and resolve the conflicts?

File: src/azul/indexer.py (hunk 2 above)
Comment (jessebrennan) [reply src/azul/indexer.py]: done

File: src/azul/indexer.py (hunk 3 above)
Comment (jessebrennan) [reply src/azul/indexer.py]: :+1:

File: src/azul/indexer.py (hunk 4 above)
Comment (jessebrennan) [reply src/azul/indexer.py]: :+1:

File: src/azul/indexer.py (hunk 5 above)
Comment (jessebrennan) [reply src/azul/indexer.py]: :+1:

File: src/azul/transformer.py (hunk 6 above)
Comment (jessebrennan) [reply src/azul/transformer.py]: changed to "is a deleted bundle" since "has been" isn't as temporally precise.

**Pull request merged**